Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp

```diff
@@ -111,19 +117,183 @@ void WebDriverBidiProcessor::sendMessageToFrontend(const String& message)
+            callback(makeUnexpected(result.error()));
+            return;
+        }
+
+        callback({ });
+    });
+}
+
+void WebDriverBidiProcessor::close(const BrowsingContext& browsingContext, std::optional<bool>&& optionalPromptUnload, CommandCallback<void>&& callback)
+{
+    RefPtr session = m_session.get();
+    if (!session)
+        ASYNC_FAIL_WITH_PREDEFINED_ERROR(InternalError);
+
+    // FIXME: implement `promptUnload` option.
+    // FIXME: raise `invalid argument` if `browsingContext` is not a top-level traversable.
+
+    session->closeBrowsingContext(browsingContext);
+
+    callback({ });
+}
+
+void WebDriverBidiProcessor::create(Inspector::Protocol::BidiBrowsingContext::CreateType requestedCreateType, const BrowsingContext& optionalReferenceContext, std::optional<bool>&& optionalBackground, const String& optionalUserContext, CommandCallback<BrowsingContext>&& callback)
+{
+    RefPtr session = m_session.get();
+    if (!session)
+        ASYNC_FAIL_WITH_PREDEFINED_ERROR(InternalError);
+
+    // FIXME: implement `referenceContext` option.
+    // FIXME: implement `background` option.
+    // FIXME: implement `userContext` option.
+
+    auto createType = Inspector::Protocol::BidiBrowsingContext::CreateType::Tab;
+    if (requestedCreateType != createType)
+        createType = Inspector::Protocol::BidiBrowsingContext::CreateType::Window;
+
+    auto presentationHint = createType == Inspector::Protocol::BidiBrowsingContext::CreateType::Tab ? Inspector::Protocol::Automation::BrowsingContextPresentation::Tab : Inspector::Protocol::Automation::BrowsingContextPresentation::Window;
+    session->createBrowsingContext(presentationHint, [callback = WTFMove(callback)](CommandResultOf<BrowsingContext, Inspector::Protocol::Automation::BrowsingContextPresentation>&& result) {
+        if (!result) {
+            callback(makeUnexpected(result.error()));
+            return;
+        }
+
+        auto [resultContext, resultPresentation] = WTFMove(result.value());
+        callback(resultContext);
+    });
+}
+
+void WebDriverBidiProcessor::getTree(const BrowsingContext& optionalRoot, std::optional<double>&& optionalMaxDepth, CommandCallback<Ref<JSON::ArrayOf<Protocol::BidiBrowsingContext::Info>>>&& callback)
+{
+    RefPtr session = m_session.get();
+    if (!session)
+        ASYNC_FAIL_WITH_PREDEFINED_ERROR(InternalError);
+
+    // FIXME: implement `root` option.
+    // FIXME: implement `maxDepth` option.
+
+    Vector<Ref<WebPageProxy>> pages;
+    for (Ref process : session->protectedProcessPool()->processes()) {
+        for (Ref page : process->pages()) {
+            if (!page->isControlledByAutomation())
+                continue;
+
+            pages.append(WTFMove(page));
+
+            // FIXME: add browsingContexts for frame tree nodes.
+        }
+    }
+
+    Ref<JSON::ArrayOf<Inspector::Protocol::BidiBrowsingContext::Info>> infos = JSON::ArrayOf<Inspector::Protocol::BidiBrowsingContext::Info>::create();
```

Comment:
why not do this all in one `for`?  is there some reason we have to have the intermediate `Vector`?  or can we just add to this JSON directly?

Reply:
The page iteration code is just a placeholder, we will want to share this code in several places and also iterate over frames. We might want to ask the client (Safari etc) to filter out prewarmed processes, extension pages, about:blank style pages, etc.

 I want to separate page and frame tree iteration and filtering, from the mechanical building up of the protocol objects. I'll make this more clear with builders/functors once the opportunity for code use arises.

Reply:
it still seems like unless we plan to change the order or something like that it would be easy to combine it all into a single `for`
```cpp
for (Ref process : session->protectedProcessPool()->processes()) {
    for (Ref page : process->pages()) {
        if (!page->isControlledByAutomation())
            continue;
        if (/* some other check */)
            continue;
        infos->addItem(...);
    }
}
```
but i suppose what you have is also fine